test/unit-tests/expression/parse.test.js

```diff
@@ -2231,4 +2241,228 @@ describe('parse', function () {
 
     assert.strictEqual(mathClone.evaluate('2'), 2)
   })
+
+  describe('sources', function () {
+    it('adds sources for constants', function () {
+      const parsed = math.parse('4')
+      assert.deepStrictEqual(parsed.sources, [{ index: 0, text: '4' }])
+    })
+
+    it('adds sources for symbols', function () {
+      const parsed = math.parse('foo')
+      assert.deepStrictEqual(parsed.sources, [{ index: 0, text: 'foo' }])
+    })
+
+    it('adds sources for operators', function () {
+      const parsed = math.parse('1 + 2')
+      assert.deepStrictEqual(parsed.sources, [{ index: 2, text: '+' }])
+      assert.deepStrictEqual(parsed.args[0].sources, [{ index: 0, text: '1' }])
+      assert.deepStrictEqual(parsed.args[1].sources, [{ index: 4, text: '2' }])
+    })
+
+    it('adds sources for blocks', function () {
+      const parsed = math.parse('1 + 1; 2 + 2\n3 + 3')
+
+      // should have a source for each block delimiter
+      const expected = [
+        { index: 5, text: ';' },
+        { index: 12, text: '\n' }
+      ]
+
+      assert.deepStrictEqual(parsed.sources, expected)
+    })
+
+    it('adds sources for 1D matrices', function () {
+      const parsed = math.parse('[1, 2, 3]')
+
+      // should have a source for brackets and item delimiters
+      const expected = [
+        { index: 0, text: '[' },
+        { index: 2, text: ',' },
+        { index: 5, text: ',' },
+        { index: 8, text: ']' }
+      ]
+
+      assert.deepStrictEqual(parsed.sources, expected)
+    })
+
+    it('adds sources for 2D matrices', function () {
+      const parsed = math.parse('[1, 2; 3, 4]')
+
+      // outer matrix has sources for brackets and row delimeters
+      const expected = [
+        { index: 0, text: '[' },
+        { index: 5, text: ';' },
```

Comment:
Shouldn't the sources include the two comma's too?

Reply:
A matrix is represented as an Array of Arrays, so the commas are mapped to the child Arrays. As a rule, I tried to make it so that every token mapped back to exactly 1 parent

Reply:
Ah, of course, never mind, I was wrongly thinking that a single ArrayNode was created with nested contents. Your approach to map 1 token to 1 node totally makes sense.